exchanges/newexchange.md

```diff
@@ -0,0 +1,766 @@
+
+https://docs.ftx.com/#fills-2
+
+```go
+// WsFills stores websocket fills' data
+type WsFills struct {
+	Fee       float64   `json:"fee"`
+	FeeRate   float64   `json:"feeRate"`
+	Future    string    `json:"future"`
+	ID        int64     `json:"id"`
+	Liquidity string    `json:"liquidity"`
+	Market    string    `json:"market"`
+	OrderID   int64     `json:"int64"`
+	TradeID   int64     `json:"tradeID"`
+	Price     float64   `json:"price"`
+	Side      string    `json:"side"`
+	Size      float64   `json:"size"`
+	Time      time.Time `json:"time"`
+	OrderType string    `json:"orderType"`
+}
+
+// WsFillsDataStore stores ws fills' data
+type WsFillsDataStore struct {
+	Channel     string  `json:"channel"`
+	MessageType string  `json:"type"`
+	FillsData   WsFills `json:"fills"`
+}
+```
+
+- Create the authentication function based on specifications provided in the documentation:
+
+https://docs.ftx.com/#private-channels
+
+```go
+// WsAuth sends an authentication message to receive auth data
+func (f *FTX) WsAuth() error {
+	intNonce := time.Now().UnixNano() / 1000000
+	strNonce := strconv.FormatInt(intNonce, 10)
+	hmac := crypto.GetHMAC(
+		crypto.HashSHA256,
+		[]byte(strNonce+"websocket_login"),
+		[]byte(f.API.Credentials.Secret),
+	)
+	sign := crypto.HexEncodeToString(hmac)
+	req := Authenticate{Operation: "login",
+		Args: AuthenticationData{
+			Key:  f.API.Credentials.Key,
+			Sign: sign,
+			Time: intNonce,
+		},
+	}
+	return f.WebsocketConn.SendJSONMessage(req)
+}
+```
+
+- Create function to generate default subscriptions
+
+```go
+// GenerateDefaultSubscriptions generates default subscription
+func (f *FTX) GenerateDefaultSubscriptions() {
+	var channels = []string{wsTicker, wsTrades, wsOrderbook, wsMarkets, wsFills, wsOrders} // All the channels that can be subscribed to
+	var subscriptions []wshandler.WebsocketChannelSubscription
+	for a := range f.CurrencyPairs.AssetTypes {
+		pairs := f.GetEnabledPairs(f.CurrencyPairs.AssetTypes[a])
+		newPair := currency.NewPairWithDelimiter(pairs[0].Base.String(), pairs[0].Quote.String(), "-")
+		for x := range channels {
+			subscriptions = append(subscriptions, wshandler.WebsocketChannelSubscription{
+				Channel:  channels[x],
+				Currency: newPair,
+			})
+		}
+	}
+	f.Websocket.SubscribeToChannels(subscriptions)
+}
+```
+
+- Create subscribe function with the data provided by the exchange documentation:
+
+https://docs.ftx.com/#request-process
+
+Create a struct required to subscribe to channels:
+
+```go
+// WsSub has the data used to subscribe to a channel
+type WsSub struct {
+	Channel   string `json:"channel,omitempty"`
+	Market    string `json:"market,omitempty"`
+	Operation string `json:"op,omitempty"`
+}
+```
+
+Create the subscription function:
+
+```go
+// Subscribe sends a websocket message to receive data from the channel
+func (f *FTX) Subscribe(channelToSubscribe wshandler.WebsocketChannelSubscription) error {
```

Comment:
Add subscribe and functionality example to websocket compatability in wrapper.go file